src/datalab_app_plugin_insitu/utils.py

```diff
@@ -318,28 +337,134 @@ def _process_data(
         nmr_dimension = check_nmr_dimension(nmr_folder_path)
 
         if nmr_dimension == "1D":
-            spec_paths, acqu_paths = setup_paths(nmr_folder_path, start_at, exclude_exp)
+            spec_paths, acqu_paths = setup_paths(
+                nmr_folder_path, start_at, exclude_exp)
             time_points = process_time_data(acqu_paths)
-            nmr_data, df, num_experiments = process_spectral_data(spec_paths, time_points)
+            nmr_data, df, num_experiments = process_spectral_data(
+                spec_paths, time_points)
 
         elif nmr_dimension == "pseudo2D":
             exp_folders = [
                 d for d in Path(nmr_folder_path).iterdir() if d.is_dir() and d.name.isdigit()
             ]
             if not exp_folders:
-                raise FileNotFoundError("No experiment folders found in NMR data")
+                raise FileNotFoundError(
+                    "No experiment folders found in NMR data")
 
             exp_folder = str(Path(nmr_folder_path) / exp_folders[0])
-            nmr_data, df, num_experiments = process_pseudo2d_spectral_data(exp_folder)
+            nmr_data, df, num_experiments = process_pseudo2d_spectral_data(
+                exp_folder)
 
         else:
             raise ValueError(f"Unknown NMR dimension type: {nmr_dimension}")
 
         merged_df = (
-            process_echem_data(base_folder, echem_folder_name) if echem_folder_name else None
+            process_echem_data(
+                base_folder, echem_folder_name) if echem_folder_name else None
         )
 
-        return prepare_for_bokeh(nmr_data, df, merged_df, num_experiments)
+        result = prepare_for_bokeh(nmr_data, df, merged_df, num_experiments)
+
+        if result is None:
+            raise RuntimeError(
+                "prepare_for_bokeh returned None instead of expected data")
+
+        return result
 
     except Exception as e:
         raise RuntimeError(f"Error in common processing: {str(e)}")
+
+
+def get_cache_path(file_id: str) -> str:
+    """Get the path to the cache file for a given file_id."""
+    cache_dir = os.path.join(os.path.expanduser("~"), ".datalab_cache")
+    os.makedirs(cache_dir, exist_ok=True)
+
+    cache_path = os.path.join(cache_dir, f"{file_id}_insitu.h5")
```

Comment:
```suggestion
    cache_dir = os.path.join(os.path.expanduser("~"), ".datalab_cache")
    os.makedirs(cache_dir, exist_ok=True)

    cache_path = os.path.join(cache_dir, f"{file_id}_insitu.h5")
```

Looks good, but lets use the folder that the file is saved in already for the cache, e.g., `file_info["location"].parent`